### DISCUSSION TITLE
npm install wrapper

### DISCUSSION THREAD

**Discussion by blaix** [proofrock/ws4sqlite#47] [Ideas]
I'd like to create a wrapper to install ws4sqlite with npm. It would just look at the platform and architecture and download the appropriate binary from the github releases page to the node modules bin path. This is to make it easier to run on my render.com sites where I'm not using Docker and orchestrating everything with npm. Any thoughts or objections on this? or better ideas? Thanks.

**Comment by proofrock** [comment]
Hi Justin! I am not very familiar with NPM so I cannot contribute much - and for sure I have no objections! Thanks for contributing, please let me know if you need anything from "my side", and how can I link/give proper attribution to you and the project, when it's done!

I will definitely look into it, it looks... enabling.

**Comment by blaix** [comment]
Cool, thanks. Before I get too far into this, curious if you've seen or have thoughts on libsql, the fork of sqlite with a [server](https://github.com/tursodatabase/libsql/tree/main/libsql-server) that also supports [sql-over-http](https://docs.turso.tech/sdk/http/reference)?

**Reply by proofrock** [reply]
Thanks for the pointer! I will be going back to it for a deeper glance, but at first impression it seems to me that ws4sqlite is a bit more featureful, or just... fills a different niche. I can't see much discussion about transactions, in their documentation; the "continuation" style is very cool (I may steal here and there a little bit ;-) ), but they don't go very deep in discussing how it relates to a single-threaded, single transaction db. It looks like an alternative access method to their existing project, which is of course their job - and they do it perfectly well. Also, it seems to be quite oriented to backend-to-backend communication.

On the other hand I tried (tried!) to bake a bit more "structure" into ws4sqlite - initialization SQL statements, cron-activated management tasks, sql stored on the server (turso/hrana has it, but I think it still must come from the client), and some features that enables you to call it from a browser in a safe way - for some use cases - as I had this in mind at the beginning. Multiple db support, a web server, values batching - also features, of course they might also be seen as overkill.

Also, there is some work to enable other database engines - which I think is cool, so it will become ws4sql, one ([near](https://github.com/proofrock/ws4sqlite/releases/tag/ws4sql_v0.17dev2)) day.

All in all, I am very happy for any competition. I think it's great to have validation of the concept, and every implementation is a bit different, in the end giving great choice. I am quite happy with ws4sqlite, and I also use it; but if someone else makes something better, I may end up using it or contributing to it, which is a win for me. In the meanwhile, I'll keep gathering ideas from users and implementing them as well as I can.

**Comment by blaix** [comment]
Personally, tying myself to a corporate entity, even if they promise to keep the tech free and open, makes me nervous, so I'm glad ws4sql[ite] exists, but still wondering how you think about this project compared to what they're doing.

**Reply by proofrock** [reply]
I am not against companies providing OSS projects; let's face it, I work on it on a percentage of my spare time, when my wife is not feeling neglected; while they can dedicate more resources, so the scope is for sure more ambitious. It's ok.  What I'm trying to do is to stay small and clean, while implementing ideas I had, that may be useful or even original. And listen to feedback, and gather new ideas and challenges. ws4sqlite won't be a big project like, say, pocketbase (or turso), but it aims to be a component of someone else's projects, and for that it needs to do - and do well - just few things. Turso adds a lot on top of SQLite, and that is quite cool. I am a bit envious ;-) because I'd like to work more on this, of course, but also I'm having fun being an "artisan", so...

Anyway: in the end it's not what I think, but what _you_ - the user - think. It's your choice to use one or the other, or even to contribute; and that's the good thing of it all. May I ask you what do you think of the two projects?

**Reply by blaix** [reply]
I am also not against companies providing OSS. I think it's really cool. But they seem to lean towards "move fast and break things" when I personally prefer "move slow and fix things". And it feels like going with turso or similar means I am building my project in that company's ecosystem, and would be subject to their whims and priorities. That's not for me. I like what you're saying about doing a few things well, and it being a component of someone else's project. I'm much happier to plug an artisan's project into my own than to plug my project into a company's product.

I'm also a fan of projects where I can communicate with the creator :)

I will admit a nice thing about going with something like turso would be the push-button, fully managed setup vs configuring and hosting it all yourself, but for me the trade-off is not worth it for the reasons stated above.

Thank you for the detailed responses, it makes me even happier to use your project!

**Reply by proofrock** [reply]
Ah, you're welcome, in every sense 😉

**Comment by blaix** [comment]
Would it be better to call the npm package `ws4sql` and target the upcoming fork?

**Reply by proofrock** [reply]
It may be a good idea. The timeframe that I gave myself is a couple of months ("August" let's say); is it something you can work with? Anyways, the prereleases that I released are "stable" so to speak, but there might be some other breaking changes (I don't anticipate any but might be).

**Reply by blaix** [reply]
"months" is always a time frame I can work with, I usually have 3-5 things going at a time :) Thanks for the context.

**Comment by blaix** [comment]
Released the first version earlier today. It targets the 0.17-dev2 release of the ws4sql fork. https://github.com/blaix/ws4sql-npm/

Would love it if anyone else could verify that it installs and works for them ok, since it has a lot of platform and cpu architecture dependent logic.

**Reply by proofrock** [reply]
Thanks Justin!! I tested it under linux, both AMD64 (Arch Linux) and AARCH64 (raspbian), it seems to work as intended.

I think it needs root access for installation (the `npm install -g` step): may it be worthwhile to specify it in the README? It's pretty obvious from the error messages, still it would be nice.

Later on, when I will release `ws4sql` (minus `-ite`) we will talk about how to integrate it into the documentation.

**Reply by proofrock** [reply]
Also tested under windows x64, works fine. Kudos! :-)
